File: app/src/main/java/com/thoughtcrimes/securesms/components/AlbumThumbnailView.java
package com.thoughtcrimes.securesms.components;

import android.content.Context;
import android.util.AttributeSet;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.TextView;

import androidx.annotation.ColorInt;
import androidx.annotation.IdRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.thoughtcrimes.securesms.mms.Slide;
import com.thoughtcrimes.securesms.mms.SlideClickListener;
import com.thoughtcrimes.securesms.mms.SlidesClickedListener;
import com.beldex.libbchat.utilities.Stub;
import com.thoughtcrimes.securesms.mms.GlideRequests;

import java.util.List;

import io.beldex.bchat.R;

public class AlbumThumbnailView extends FrameLayout {

  private @Nullable
  SlideClickListener thumbnailClickListener;
  private @Nullable
  SlidesClickedListener downloadClickListener;

  private int currentSizeClass;

  private ViewGroup                 albumCellContainer;
  private Stub<TransferControlView> transferControls;

  private final SlideClickListener defaultThumbnailClickListener = (v, slide) -> {
      if (thumbnailClickListener != null) {
        thumbnailClickListener.onClick(v, slide);
      }
  };

  private final OnLongClickListener defaultLongClickListener = v -> this.performLongClick();

  public AlbumThumbnailView(@NonNull Context context) {
    super(context);
    initialize();
  }

  public AlbumThumbnailView(@NonNull Context context, @Nullable AttributeSet attrs) {
    super(context, attrs);
    initialize();
  }

  private void initialize() {
    inflate(getContext(), R.layout.album_thumbnail_view, this);

    albumCellContainer = findViewById(R.id.albumCellContainer);
    transferControls   = new Stub<>(findViewById(R.id.albumTransferControlsStub));
  }

  public void setSlides(@NonNull GlideRequests glideRequests, @NonNull List<Slide> slides, boolean showControls) {
    if (slides.size() < 2) {
      throw new IllegalStateException("Provided less than two slides.");
    }

    if (showControls) {
      transferControls.get().setShowDownloadText(true);
      transferControls.get().setSlides(slides);
      transferControls.get().setDownloadClickListener(v -> {
        if (downloadClickListener != null) {
          downloadClickListener.onClick(v, slides);
        }
      });
    } else {
      if (transferControls.resolved()) {
        transferControls.get().setVisibility(GONE);
      }
    }

    int sizeClass = Math.min(slides.size(), 6);

    if (sizeClass != currentSizeClass) {
      inflateLayout(sizeClass);
      currentSizeClass = sizeClass;
    }

    showSlides(glideRequests, slides);
  }

  public void setCellBackgroundColor(@ColorInt int color) {
    ViewGroup cellRoot = findViewById(R.id.album_thumbnail_root);

    if (cellRoot != null) {
      for (int i = 0; i < cellRoot.getChildCount(); i++) {
        cellRoot.getChildAt(i).setBackgroundColor(color);
      }
    }
  }

  public void setThumbnailClickListener(@Nullable SlideClickListener listener) {
    thumbnailClickListener = listener;
  }

  public void setDownloadClickListener(@Nullable SlidesClickedListener listener) {
    downloadClickListener = listener;
  }

  private void inflateLayout(int sizeClass) {
    albumCellContainer.removeAllViews();

    switch (sizeClass) {
      case 2:
        inflate(getContext(), R.layout.album_thumbnail_2, albumCellContainer);
        break;
      case 3:
        inflate(getContext(), R.layout.album_thumbnail_3, albumCellContainer);
        break;
      case 4:
        inflate(getContext(), R.layout.album_thumbnail_4, albumCellContainer);
        break;
      case 5:
        inflate(getContext(), R.layout.album_thumbnail_5, albumCellContainer);
        break;
      default:
        inflate(getContext(), R.layout.album_thumbnail_many, albumCellContainer);
        break;
    }
  }

  private void showSlides(@NonNull GlideRequests glideRequests, @NonNull List<Slide> slides) {
    setSlide(glideRequests, slides.get(0), R.id.album_cell_1);
    setSlide(glideRequests, slides.get(1), R.id.album_cell_2);

    if (slides.size() >= 3) {
      setSlide(glideRequests, slides.get(2), R.id.album_cell_3);
    }

    if (slides.size() >= 4) {
      setSlide(glideRequests, slides.get(3), R.id.album_cell_4);
    }

    if (slides.size() >= 5) {
      setSlide(glideRequests, slides.get(4), R.id.album_cell_5);
    }

    if (slides.size() > 5) {
      TextView text = findViewById(R.id.album_cell_overflow_text);
      text.setText(getContext().getString(R.string.AlbumThumbnailView_plus, slides.size() - 5));
    }
  }

  private void setSlide(@NonNull GlideRequests glideRequests, @NonNull Slide slide, @IdRes int id) {
  }
}

File: app/src/main/java/io/beldex/bchat/conversation/v2/components/AlbumThumbnailView.kt
package io.beldex.bchat.conversation.v2.components

import android.content.Context
import android.graphics.Canvas
import android.graphics.Rect
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.ViewGroup
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.core.view.children
import androidx.core.view.isVisible
import com.beldex.libbchat.messaging.sending_receiving.attachments.AttachmentTransferProgress
import com.beldex.libbchat.messaging.sending_receiving.attachments.DatabaseAttachment
import com.beldex.libbchat.utilities.recipients.Recipient
import io.beldex.bchat.MediaPreviewActivity
import io.beldex.bchat.components.CornerMask
import io.beldex.bchat.conversation.v2.utilities.ThumbnailView
import io.beldex.bchat.database.model.MmsMessageRecord
import com.bumptech.glide.RequestManager
import io.beldex.bchat.mms.Slide
import io.beldex.bchat.util.ActivityDispatcher
import io.beldex.bchat.R
import io.beldex.bchat.conversation.v2.messages.VisibleMessageContentView
import io.beldex.bchat.databinding.AlbumThumbnailViewBinding
import io.beldex.bchat.util.DateUtils
import java.util.Locale

class AlbumThumbnailView : RelativeLayout {

    companion object {
        const val MAX_ALBUM_DISPLAY_SIZE = 3
    }

    private val binding: AlbumThumbnailViewBinding by lazy { AlbumThumbnailViewBinding.bind(this) }

    // region Lifecycle
    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    private val cornerMask by lazy {
        CornerMask(
            this
        )
    }
    private var slides: List<Slide> = listOf()
    private var slideSize: Int = 0


    override fun dispatchDraw(canvas: Canvas) {
        super.dispatchDraw(canvas)
        cornerMask.mask(canvas)
    }
    // endregion

    // region Interaction

    fun calculateHitObject(event: MotionEvent, mms: MmsMessageRecord, threadRecipient: Recipient,  onAttachmentNeedsDownload: (Long, Long) -> Unit) {
        val rawXInt = event.rawX.toInt()
        val rawYInt = event.rawY.toInt()
        val eventRect = Rect(rawXInt, rawYInt, rawXInt, rawYInt)
        val testRect = Rect()
        // test each album child
        binding.albumCellContainer.findViewById<ViewGroup>(R.id.album_thumbnail_root)?.children?.forEachIndexed forEach@{ index, child ->
            child.getGlobalVisibleRect(testRect)
            if (testRect.contains(eventRect)) {
                // hit intersects with this particular child
                val slide = slides.getOrNull(index) ?: return@forEach
                // only open to downloaded images
                if (slide.transferState == AttachmentTransferProgress.TRANSFER_PROGRESS_FAILED) {
                    // Restart download here (on IO thread)
                    (slide.asAttachment() as? DatabaseAttachment)?.let { attachment ->
                        onAttachmentNeedsDownload(attachment.attachmentId.rowId, mms.getId())
                    }
                }
                if (slide.isInProgress) return@forEach

                ActivityDispatcher.get(context)?.dispatchIntent { context ->
                    MediaPreviewActivity.getPreviewIntent(context, slide, mms, threadRecipient)
                }
            }
        }
    }

    fun clearViews() {
        binding.albumCellContainer.removeAllViews()
        slideSize = -1
    }

    fun bind(glideRequests: RequestManager, message: MmsMessageRecord,
             isStart: Boolean, isEnd: Boolean) {
        slides = message.slideDeck.thumbnailSlides
        if (slides.isEmpty()) {
            // this should never be encountered because it's checked by parent
            return
        }
        //calculateRadius(isStart, isEnd, message.isOutgoing)

        // recreate cell views if different size to what we have already (for recycling)
        if (slides.size != this.slideSize) {
            binding.albumCellContainer.removeAllViews()
            LayoutInflater.from(context).inflate(layoutRes(slides.size), binding.albumCellContainer)
            val overflowed = slides.size > MAX_ALBUM_DISPLAY_SIZE
            binding.albumCellContainer.findViewById<TextView>(R.id.album_cell_overflow_text)?.let { overflowText ->
                // overflowText will be null if !overflowed
                overflowText.isVisible = overflowed // more than max album size
                overflowText.text = context.getString(R.string.AlbumThumbnailView_plus, slides.size - MAX_ALBUM_DISPLAY_SIZE)
            }
            this.slideSize = slides.size
        }
        // iterate binding
        slides.take(MAX_ALBUM_DISPLAY_SIZE).forEachIndexed { position, slide ->
            val thumbnailView = getThumbnailView(position)
            thumbnailView.setImageResource(glideRequests, slide, isPreview = false, mms = message)
        }
    }

    // endregion


    fun layoutRes(slideCount: Int) = when (slideCount) {
        1 -> R.layout.album_thumbnail_1 // single
        2 -> R.layout.album_thumbnail_2// two sidebyside
        else -> R.layout.album_thumbnail_3 // three stacked with additional text
    }

    fun getThumbnailView(position: Int): ThumbnailView = when (position) {
        0 -> binding.albumCellContainer.findViewById<ViewGroup>(R.id.albumCellContainer).findViewById(R.id.album_cell_1)
        1 -> binding.albumCellContainer.findViewById<ViewGroup>(R.id.albumCellContainer).findViewById(R.id.album_cell_2)
        2 -> binding.albumCellContainer.findViewById<ViewGroup>(R.id.albumCellContainer).findViewById(R.id.album_cell_3)
        else -> throw Exception("Can't get thumbnail view for non-existent thumbnail at position: $position")
    }

    fun calculateRadius(isStart: Boolean, isEnd: Boolean, outgoing: Boolean) {
        val roundedDimen = context.resources.getDimension(R.dimen.message_corner_radius).toInt()
        val collapsedDimen = context.resources.getDimension(R.dimen.message_corner_collapse_radius).toInt()
        val (startTop, endTop, startBottom, endBottom) = when {
            // single message, consistent dimen
            isStart && isEnd -> intArrayOf(roundedDimen, roundedDimen, roundedDimen, roundedDimen)
            // start of message cluster, collapsed BL
            isStart -> intArrayOf(roundedDimen, roundedDimen, collapsedDimen, roundedDimen)
            // end of message cluster, collapsed TL
            isEnd -> intArrayOf(collapsedDimen, roundedDimen, roundedDimen, roundedDimen)
            // else in the middle, no rounding left side
            else -> intArrayOf(collapsedDimen, roundedDimen, collapsedDimen, roundedDimen)
        }
        // TL, TR, BR, BL (CW direction)
        cornerMask.setRadii(
                if (!outgoing) startTop else endTop, // TL
                if (!outgoing) endTop else startTop, // TR
                if (!outgoing) endBottom else startBottom, // BR
                if (!outgoing) startBottom else endBottom // BL
        )
    }

}